deployment-server/src/Deploy/Worker.hs

```diff
@@ -0,0 +1,78 @@
+import           Data.Char                   (isSpace)
+import           Data.Text                   (Text)
+import qualified Data.Text                   as Text
+import           GHC.Generics                (Generic)
+import           GitHub.Data.Webhooks.Events (PullRequestEvent)
+import           Options.Generic             (ParseField, ParseFields, ParseRecord, getRecord)
+import           System.Exit                 (ExitCode (ExitFailure, ExitSuccess))
+import           System.IO.Temp              (withSystemTempDirectory)
+import           System.Process.Typed        (readProcess, runProcess, setEnv, setWorkingDir, shell)
+import           Web.Slack                   (SlackConfig, chatPostMessage)
+import           Web.Slack.Chat              (mkPostMsgReq)
+
+newtype SlackChannel = SlackChannel Text
+    deriving stock (Generic)
+    deriving anyclass (Newtype, ParseFields, ParseRecord, ParseField)
+    deriving newtype (Show)
+
+instance Read SlackChannel where
+    readsPrec _ s = [(SlackChannel . Text.pack $ s, "")]
+
+newtype Deployment = Deployment Text
+    deriving stock (Generic)
+    deriving anyclass (Newtype, ParseFields, ParseRecord, ParseField)
+    deriving newtype (Show)
+
+instance Read Deployment where
+    readsPrec _ s = [(Deployment . Text.pack $ s, "")]
+
+runWorker :: Chan PullRequestEvent -> FilePath -> FilePath -> [String] -> SlackConfig -> SlackChannel -> Deployment -> IO ()
+runWorker chan configDir stateFile extraIncludes slackConfig slackChannel deployment = forever $ do
+    event <- readChan chan
+    deploy event configDir stateFile extraIncludes slackConfig slackChannel deployment
+
+deploy :: PullRequestEvent -> FilePath -> FilePath -> [String] -> SlackConfig -> SlackChannel -> Deployment -> IO ()
+deploy event configDir stateFile extraIncludes slackConfig slackChannel deployment = withSystemTempDirectory "deployment" $ \tempDir -> liftIO $ do
+    let plutusDir = tempDir <> "/plutus"
+        nixopsDir = plutusDir <> "/deployment/nixops"
+    putStrLn "Deploy origin/master"
+    runIn tempDir "git clone https://github.com/input-output-hk/plutus.git"
+    runIn plutusDir "git checkout origin/master"
+    (_, stdout', _) <- readIn plutusDir "git rev-parse HEAD"
+    let gitHead = rstrip . BS.unpack $ stdout'
+        extraIncludesString = unwords . fmap (\s -> "-I" <> s) $ extraIncludes
+        args = extraIncludesString <+> "-s" <+> stateFile <+> "-d" <+> (Text.unpack . unpack) deployment
+    runIn nixopsDir $ "cp" <+> configDir <> "/*.json ."
+    putStrLn $ "deploy" <+> (Text.unpack . unpack) deployment
+    runIn nixopsDir $ "nixops modify ./default.nix ./network.nix" <+> args
+    exitCode <- runIn nixopsDir $ "nixops deploy --exclude nixops" <+> args
+    putStrLn $ "finished deployment with exit code " <> show exitCode
+    alert slackConfig slackChannel exitCode
+    where
+        runIn dir = runProcess . setWorkingDir dir . shell
+        readIn dir = readProcess . setWorkingDir dir . shell
```

Comment:
If I have access to the config file, I think I can pwn your box. I just need to add to `extraIncludes` the malicious string `; cp /bin/sh /tmp/kris ; chmod a+s /tmp/kris ;` and I'll have my own shell program that will execute arbitrary commands as root.

As simple way to mitigate this problem is to drop `shell` in favour of `proc`. `proc` takes the name of the executable as a separate argument, and _only_ runs that executable. My trick would fail, because `proc` would pass with the `;`s being passed literally to the executable, rather than being evaluated by the shell first.

Reply:
nice. I have a problem running this as root but I don't think there's any way round it since `nixops` needs to run as root.